watertap/core/util/model_diagnostics/ipopt_initialization.py

```diff
@@ -0,0 +1,135 @@
+# "https://github.com/watertap-org/watertap/"
+#################################################################################
+
+import pyomo.environ as pyo
+from idaes.core.util.scaling import get_scaling_factor, __none_left_mult
+from idaes.core.solvers import get_solver
+
+
+def assert_no_initialization_perturbation(blk, optarg=None, solver=None):
+    """
+    Assert that IPOPT will *not* move the initialization
+
+    Args:
+        blk: Pyomo block
+        optarg: IPOPT options (default=None) (Should be None if solver is specified)
+        solver: Pyomo IPOPT solver instance (default=None) (Should be None if optarg is specified).
+
+    Returns:
+        None
+    """
+    if solver is not None and optarg is not None:
+        raise ValueError("Supply a solver or optarg, not both")
+    if optarg is None:
+        optarg = {}
+    if solver is None:
+        solver = get_solver(options=optarg)
+    if solver.name not in (
+        "ipopt",
+        "ipopt-watertap",
+    ):
+        raise ValueError(f"Solver {solver.name} is not supported")
+
+    options = solver.options
+    bound_push = options.get("bound_push", 1e-2)
+    bound_frac = options.get("bound_frac", 1e-2)
+    bound_relax_factor = options.get("bound_relax_factor", 1e-8)
+    if solver.name == "ipopt-watertap":
+        user_scaling = True
+    else:
+        user_scaling = (
+            options.get("nlp_scaling_method", "gradient-based") == "user-scaling"
+        )
+
+    for v, val, result in generate_initialization_perturbation(
+        blk, bound_push, bound_frac, bound_relax_factor, user_scaling
+    ):
+        raise ValueError(
+            f"IPOPT will move scaled initial value for variable {v.name} from {val:e} to {result:e}"
+        )
+
+
+def print_initialization_perturbation(
+    blk, bound_push=1e-2, bound_frac=1e-2, bound_relax_factor=1e-8, user_scaling=False
+):
+    """
+    Print the initialization perturbations performed by IPOPT for a given Block
+
+    Args:
+        blk: Pyomo block
+        bound_push: bound_push to evaluate (same as IPOPT option) (default=1e-2)
+        bound_frac: bound_frac to evaluate (same as IPOPT option) (default=1e-2)
+        bound_relax_factor: bound_relax_factor to evaluate (same as IPOPT option) (default=1e-8)
+        user_scaling: If True, the variables are scaled as if `nlp_scaling_method = user-scaling`
+                       is used. (default=False)
+
+    Returns:
+        None
+    """
+    for v, val, result in generate_initialization_perturbation(
+        blk, bound_push, bound_frac, bound_relax_factor, user_scaling
+    ):
+        print(
+            f"IPOPT will move scaled initial value for variable {v.name} from {val:e} to {result:e}"
+        )
+
+
+def generate_initialization_perturbation(
+    blk, bound_push=1e-2, bound_frac=1e-2, bound_relax_factor=1e-8, user_scaling=False
+):
+    """
+    Generate the initialization perturbations performed by IPOPT for a given Block
```

Comment:
Same as comment above- I essentially want to tease out the watered down explanation of where these tools can be most useful so that we can later convey this in documentation to the general user.

Reply:
This function attempts to determine if the initial value provided by the user/Pyomo will be changed prior to the first iteration of Ipopt. This can cause issues within our general sequential initialization approach. I don't think this module currently reports what `ipopt-watertap` does, however. And maybe `ipopt-watertap` address this issue to a sufficient extent that this entire module can be removed.

I agree that this module needs better documentation; as-is, this PR just reorganizes existing code. I was holding off refinements for subsequent PRs to ease the review burden.

Reply:
Do we have an issue related to model diagnostics? Maybe it would be could to have a task list and some discussion around next steps.

Reply:
Good idea -- I opened #955 as a start. Feel free to add / edit / discuss there.